internal/provider/data_source_gitlab_current_user.go

```diff
@@ -0,0 +1,101 @@
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_current_user`" + ` data source allows details of the current user (determined by the input provider token) to be retrieved.
+
+**Upstream API**: [GitLab REST API docs](https://docs.gitlab.com/ee/api/graphql/reference/index.html#querycurrentuser)`,
+
+		ReadContext: dataSourceGitlabCurrentUserRead,
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Description: "ID of the user. This is in the form of a GraphQL globally unique ID.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"username": {
+				Description: "Username of the user. Unique within this instance of GitLab.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"name": {
+				Description: "Human-readable name of the user. Returns **** if the user is a project bot and the requester does not have permission to view the project. ",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"bot": {
+				Description: "Indicates if the user is a bot.",
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+			"group_count": {
+				Description: "Group count for the user.",
+				Type:        schema.TypeInt,
+				Computed:    true,
+			},
+			"namespace_id": {
+				Description: "Personal namespace of the user. This is in the form of a GraphQL globally unique ID.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"public_email": {
+				Description: "User’s public email.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+})
+
+func dataSourceGitlabCurrentUserRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	query := GraphQLQuery{
+		`query {currentUser {name, bot, groupCount, id, namespace{id} publicEmail, username}}`,
+	}
+	log.Printf("[DEBUG] executing GraphQL Query %s to retrieve current user", query.Query)
+
+	var response CurrentUserResponse
+	_, err := SendGraphQLRequest(context.Background(), client, query, &response)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(response.Data.CurrentUser.ID)
+	d.Set("username", response.Data.CurrentUser.Username)
+	d.Set("name", response.Data.CurrentUser.Name)
+	d.Set("bot", response.Data.CurrentUser.Bot)
+	d.Set("group_count", response.Data.CurrentUser.GroupCount)
+	d.Set("namespace_id", response.Data.CurrentUser.Namespace.ID)
+	d.Set("public_email", response.Data.CurrentUser.PublicEmail)
+
+	return nil
+}
+
+// Struct representing current user based on the input API token
+type CurrentUserResponse struct {
+	Data struct {
+		CurrentUser GraphQLUser `json:"currentUser"`
+	} `json:"data"`
+}
+
+type GraphQLUser struct {
+	Name       string `json:"name"`
+	Bot        bool   `json:"bot"`
+	GroupCount int    `json:"groupCount"`
+	ID         string `json:"id"` // This is purposefully a string, as in some APIs it comes back as a globally unique ID
```

Comment:
Out of curiosity: what's a `globally unique ID` and how would the non-integer representation look like?

Reply:
very interesting! 

> I do suspect that in some cases since we have a mixture of GraphQL and REST we may have to post-process out the unique ID and store it in a separate spot for folks to use in downstream resources.

That's exactly my doubt here - we would be able to e.g. use the `namespace_id` field to create a project for a specific user ...

I need to do a little bit of research about those ids and the format. Let's see if for now and to consistently allow interoperability we could just use the last part of the `globally unique id` when splitting by `/`. 
WDYT?

Reply:
@PatrickRice-KSC I've added some logic to make this data source more interoperable with the other data sources and resources by separating the `id` from the `global_id`.

From what I could gather always the part after the `/` is an integer id mapping the normal internal id we are used to from the REST API. 

WDYT? Please feel free to change it in the way you like it!

Reply:
@timofurrer  - Makes complete sense to me, and I agree with keeping the base attribute as the integer and pre-pending the GraphQL ID with "global". I suspect that's something we should standardize across GraphQL resources (because almost every ID that comes back via GraphQL is in the global format so far that I've seen).